### PR TITLE
refactor AR install to support params, plus help manpage

### PR DESCRIPTION
**Pull request by aweiteka**

### CODE REVIEW

File: examples/atomic-registry/systemd/install.sh [hunk 1]
```diff
@@ -1,66 +1,109 @@
-#!/bin/bash
+#!/usr/bin/env bash
```
Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: This is a little surprising -- why the change?

File: examples/atomic-registry/systemd/install.sh [hunk 2]
```diff
@@ -1,66 +1,109 @@
...
+
+set -o pipefail
```
Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: Not sure if they're set elsewhere, but in general we want our Bash scripts to run with:
```
set -o errexit
set -o nounset
set -o pipefail
```

An easy way to get this is to `source hack/lib/init.sh`, but I'm not sure if that's appropriate for this script.

File: examples/atomic-registry/systemd/install.sh [hunk 3]
```diff
@@ -1,66 +1,109 @@
...
 
 INSTALL_HOST=${1:-`hostname`}
```
Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: Backtick expressions are deprecated, prefer:
```sh
INSTALL_HOST="${1:-"$(hostname)"}"
```

File: examples/atomic-registry/systemd/install.sh [hunk 4]
```diff
@@ -1,66 +1,109 @@
...
+
+# Defaults
+REGISTRYPORT=${REGISTRYPORT:-5000}
```
Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: We should quote all of our expansions:
```sh
REGISTRYPORT="${REGISTRYPORT:-5000}"
             ^                     ^
```

Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: It is custom when internalizing and defaulting values to use `$varname` lower case for script-local variables, keeping `$VARNAME` upper case for external or exported variables. See [`hack/test-go.sh`](https://github.com/openshift/origin/blob/master/hack/test-go.sh#L70-L79)

File: examples/atomic-registry/systemd/install.sh [hunk 5]
```diff
@@ -1,66 +1,109 @@
...
-echo "and/or add certificates to /etc/atomic-registry/master,"
-echo "then enable and start services:"
-echo "   sudo systemctl enable --now atomic-registry-master.service"
-echo "Once all 3 containers are running (docker ps), run the setup script"
-echo "(you can run it again if it is run early and fails)"
-echo "   sudo /var/run/setup-atomic-registry.sh ${INSTALL_HOST}"
+function write_config {
+  openshift start master --write-config /etc/atomic-registry/master \
+    --etcd-dir /var/lib/atomic-registry/etcd \
+    --public-master ${INSTALL_HOST}:${MASTERPORT} \
```
Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: Quote expansions:
```sh
    --public-master "${INSTALL_HOST}:${MASTERPORT}" \
                    ^                             ^
```

File: examples/atomic-registry/systemd/install.sh [hunk 6]
```diff
@@ -1,66 +1,109 @@
...
-sed -i 's/  templates: null$/  templates:\n    login: site\/registry-login-template.html/' /etc/atomic-registry/master/master-config.yaml
-
-echo "Optionally edit configuration file authentication /etc/atomic-registry/master/master-config.yaml,"
-echo "and/or add certificates to /etc/atomic-registry/master,"
-echo "then enable and start services:"
-echo "   sudo systemctl enable --now atomic-registry-master.service"
-echo "Once all 3 containers are running (docker ps), run the setup script"
-echo "(you can run it again if it is run early and fails)"
-echo "   sudo /var/run/setup-atomic-registry.sh ${INSTALL_HOST}"
+function write_config {
```
Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: There is a very similar function in `hack/common.sh` -- can we use that here?

File: examples/atomic-registry/systemd/install.sh [hunk 7]
```diff
@@ -1,66 +1,109 @@
...
+  openshift start master --write-config /etc/atomic-registry/master \
+    --etcd-dir /var/lib/atomic-registry/etcd \
+    --public-master ${INSTALL_HOST}:${MASTERPORT} \
+    --master https://localhost:${MASTERPORT} \
+    --listen="https://0.0.0.0:${MASTERPORT}" \
+    --cors-allowed-origins=${INSTALL_HOST}:${CONSOLEPORT}
+}
+
+function copy_files_to_host {
```
Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: Prefer to denote functions with parentheses:
```sh
function my_function_name() {
    # ....
}
```

File: examples/atomic-registry/systemd/install.sh [hunk 8]
```diff
@@ -1,66 +1,109 @@
...
+
+  echo "This directory stores the docker/distribution registry configuration file. To secure the service add TLS certificates here and reference them as environment variables." \
+      >> /etc/atomic-registry/registry/README
+  echo "This directory stores configuration and certificates for the API master." \
+      >> /etc/atomic-registry/master/README
+}
+
+function customize_config {
+  echo "Update custom ports, images and tags"
+  set -x
```
Comment (stevekuznetsov) [inline examples/atomic-registry/systemd/install.sh]: Printing the contents of the files at the end of the modifications may be less noisy than using `set -o xtrace`:
```sh
for file in /host/etc/sysconfig/atomic*; do
    echo $'\t'"${file}:"
    cat "${file}"
    echo
fi
```

**Comment by aweiteka** [comment]
cc @thomasmckay

File: examples/atomic-registry/systemd/install.sh (hunk 6 above)
Comment (aweiteka) [inline examples/atomic-registry/systemd/install.sh]: This is in a docker build context so adding a file from a parent dir is not possible unless we required the docker build context to be the entire openshift repo, which is really ugly.

File: examples/atomic-registry/systemd/install.sh (hunk 4 above)
Comment (aweiteka) [inline examples/atomic-registry/systemd/install.sh]: ACK

File: examples/atomic-registry/Makefile [hunk 9]
```diff
@@ -21,7 +21,7 @@ install-atomic:
 install-systemd:
 	atomic install $(TEST_IMAGE) $(INSTALLHOST)
 	systemctl enable --now atomic-registry-master.service
-	sleep 10
+	sleep 30
```
Comment (stevekuznetsov) [inline examples/atomic-registry/Makefile]: In a perfect world, this could be replaced with a polling loop -- although, I thought that `systemctl enable` would block until the service was running?

Comment (aweiteka) [inline examples/atomic-registry/Makefile]: Yes, polling would be ideal. The issue is I exec docker run which exits immediately. The API may not be up for a bit. We may be able to optimize and automate the setup entirely using systemd. I'd rather not dive into that for this PR.

Comment (stevekuznetsov) [inline examples/atomic-registry/Makefile]: If you're doing a `docker run` you can marry a `ExecStartPre="docker run -d container"` with `ExecStart="docker logs -f container"`. Or you may be able to do a `docker inspect` with a template to wait until something specific in the Docker API. See discussion [here](https://github.com/docker/docker/issues/6791#issuecomment-72338100) -- bumping a sleep seems like a poor way out, but I'll defer to you if you think it's out of scope for this PR.

File: examples/atomic-registry/Makefile [hunk 10]
```diff
@@ -21,8 +21,9 @@ install-atomic:
 install-systemd:
 	atomic install $(TEST_IMAGE) $(INSTALLHOST)
 	systemctl enable --now atomic-registry-master.service
-	sleep 10
-	/var/run/setup-atomic-registry.sh $(INSTALLHOST)
+	echo "Waiting for API"
+	while [[ ! $$(/bin/curl -kfLs https://localhost:8443/healthz) = "ok" ]]; do /bin/sleep 1 && echo -n "."; done
```
Comment (stevekuznetsov) [inline examples/atomic-registry/Makefile]: You don't actually need to do string comparisons since you're using the `--fail|-f` with `curl`, the result is a little more legible:

```sh
while ! /bin/curl -kfLs https://localhost:8443/healthz; do
    /bin/sleep 1
    echo -n '.'
done
```

**Comment by stevekuznetsov** [comment]
Bash bits LGTM

**Comment by aweiteka** [comment]
@smarterclayton okay to merge?

**Comment by smarterclayton** [comment]
Yes, can you squash

**Comment by aweiteka** [comment]
squashed

**Comment by aweiteka** [comment]
@smarterclayton merge?

**Comment by smarterclayton** [comment]
We're currently holding master for 1.3.0 - should be soon.

On Sep 14, 2016, at 9:52 AM, Aaron Weitekamp <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton> merge?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/10706#issuecomment-247019420>,
or mute
the thread

**Comment by smarterclayton** [comment]
LGTM [merge] thanks for your patience.
